TensorLib/NonEmptyList.lean

```diff
@@ -0,0 +1,118 @@
+import Mathlib.Tactic
+
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jean-Baptiste Tristan, Paul Govereau, Sean McLaughlin
+-/
+namespace TensorLib
+
+variable {a: Type}
+
+structure NonEmptyList (a: Type) where
+  hd: a
+  tl: List a
+  deriving BEq, Inhabited
+
+namespace NonEmptyList
+
+def toList (x: NonEmptyList a): List a := x.hd :: x.tl
+
+-- Use the nice list bracket concrete syntax
+instance NelRepr [Repr a] : Repr (NonEmptyList a) where
+  reprPrec x n := reprPrec x.toList n
+
+def length (x: NonEmptyList a): Nat := 1 + x.tl.length
+
+@[simp]
+def append (x : NonEmptyList a) (y : NonEmptyList a) : NonEmptyList a :=
+  NonEmptyList.mk x.hd (x.tl ++ y.toList)
+
+theorem appendLength (x y : NonEmptyList a) : (append x y).length = x.length + y.length := by
+  simp [length, toList]
+  linarith
+
+instance : HAppend (NonEmptyList a) (NonEmptyList a) (NonEmptyList a) where
+  hAppend := append
+
+theorem hAppendLength (x y : NonEmptyList a) : (x ++ y).length = x.length + y.length := by
+  apply appendLength
+
+@[simp]
+def appendListL (x : List a) (y : NonEmptyList a) := match x with
+| [] => y
+| x :: xs => NonEmptyList.mk x (xs ++ y.toList)
+
+theorem appendListLLength (x : List a) (y : NonEmptyList a) : (appendListL x y).length = x.length + y.length := by
+  induction x
+  . simp
+  . simp [NonEmptyList.length, NonEmptyList.toList]
+    linarith
+
+instance HAppendListL : HAppend (List a) (NonEmptyList a) (NonEmptyList a) where
+  hAppend := appendListL
+
+theorem hAppendListLLength (x : List a) (y : NonEmptyList a) : (x ++ y).length = x.length + y.length := by
+  apply appendListLLength
+
+@[simp]
+def appendListR (x : NonEmptyList a) (y : List a) := NonEmptyList.mk x.hd (x.tl ++ y)
+
+instance HAppendListR : HAppend (NonEmptyList a) (List a) (NonEmptyList a) where
+  hAppend := appendListR
+
+theorem appendListRLength (x : NonEmptyList a) (y : List a) : (appendListR x y).length = x.length + y.length := by
+  simp [NonEmptyList.length]
+  linarith
+
+theorem hAppendListRLength (x : NonEmptyList a) (y : List a) : (x ++ y).length = x.length + y.length := by
+  apply appendListRLength
+
+def all (x: NonEmptyList a) (P: a -> Bool): Bool :=
```

Comment:
Could we use coercions to simplify our code?

Reply:
Do you mean converting `NonEmptyList a` to `List a` and using `List` operations instead of duplicating the functionality?

Reply:
Wow, coercions made the example code a lot nicer. Sold.